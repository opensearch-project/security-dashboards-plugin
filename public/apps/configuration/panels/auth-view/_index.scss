.instruction-text {
  max-width: 800px;
  margin: auto;
}
