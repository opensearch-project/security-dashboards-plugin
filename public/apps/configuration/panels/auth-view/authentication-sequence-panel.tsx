/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState } from 'react';
import { EuiInMemoryTable, EuiSearchBarProps, Query } from '@elastic/eui';
import { keys, map, get } from 'lodash';
import { PanelWithHeader } from '../../utils/panel-with-header';
import { renderExpression } from '../../utils/display-utils';
import { showTableStatusMessage } from '../../utils/loading-spinner-utils';
import { DocLinks } from '../../constants';

const columns = [
  {
    field: 'order',
    name: 'Execution order',
    sortable: true,
  },
  {
    field: 'domain_name',
    name: 'Domain name',
  },
  {
    field: 'http_enabled',
    name: 'HTTP',
  },
  {
    field: 'transport_enabled',
    name: 'TRANSPORT',
  },
  {
    field: 'http_type',
    name: 'HTTP type',
  },
  {
    field: 'http_challenge',
    name: 'HTTP challenge',
  },
  {
    field: 'http_configuration',
    name: 'HTTP configuration',
    render: (config: object) => renderExpression('HTTP configuration', config),
  },
  {
    field: 'backend_type',
    name: 'Backend type',
  },
  {
    field: 'backend_configuration',
    name: 'Backend configuration',
    render: (config: object) => renderExpression('Backend configuration', config),
  },
];

const ENABLED_STRING = 'Enabled';
const DISABLED_STRING = 'Disabled';
const TRUE_STRING = 'True';
const FALSE_STRING = 'False';

export function AuthenticationSequencePanel(props: { authc: []; loading: boolean }) {
  const [query, setQuery] = useState<Query | null>(null);
  const domains = keys(props.authc);

  const items = map(domains, function (domain: string) {
    const data = get(props.authc, domain);
    // @ts-ignore
    const httpAuthenticator = data.http_authenticator;
    // @ts-ignore
    const backend = data.authentication_backend;
    return {
      // @ts-ignore
      order: data.order,
      domain_name: domain,
      // @ts-ignore
      http_enabled: data.http_enabled ? ENABLED_STRING : DISABLED_STRING,
      // @ts-ignore
      transport_enabled: data.transport_enabled ? ENABLED_STRING : DISABLED_STRING,
      http_type: httpAuthenticator.type,
      http_challenge: httpAuthenticator.challenge ? TRUE_STRING : FALSE_STRING,
      http_configuration: httpAuthenticator.config,
      backend_type: backend.type,
      backend_configuration: backend.config,
    };
  });

  const search: EuiSearchBarProps = {
    box: {
      placeholder: 'Search authentication domain',
    },
    onChange: (arg) => {
      setQuery(arg.query);
      return true;
    },
  };

  const headerText = 'Authentication sequences';

  return (
    <PanelWithHeader
      headerText={headerText}
      headerSubText="An authentication module specifies where to get the user credentials from, and against which
      backend they should be authenticated. When there are multiple authentication domains, the plugin will authenticate
      the user sequentially against each backend until one succeeds."
      helpLink={DocLinks.BackendConfigurationAuthenticationDoc}
      count={Query.execute(query || '', items).length}
    >
      <EuiInMemoryTable
        tableLayout={'auto'}
        columns={columns}
        items={items}
        itemId={'domain_name'}
        pagination={true}
        sorting={{ sort: { field: 'order', direction: 'desc' } }}
        search={search}
        message={showTableStatusMessage(props.loading, items)}
      />
    </PanelWithHeader>
  );
}
