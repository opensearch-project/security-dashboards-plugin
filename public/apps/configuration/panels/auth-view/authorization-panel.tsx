/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useState } from 'react';
import { EuiInMemoryTable, EuiEmptyPrompt, EuiSearchBarProps, Query } from '@elastic/eui';
import { keys, map, get } from 'lodash';
import { ClientConfigType } from '../../../../types';
import { PanelWithHeader } from '../../utils/panel-with-header';
import { renderExpression, ExternalLinkButton } from '../../utils/display-utils';
import { showTableStatusMessage } from '../../utils/loading-spinner-utils';
import { DocLinks } from '../../constants';

const columns = [
  {
    field: 'domain_name',
    name: 'Domain name',
  },
  {
    field: 'http_enabled',
    name: 'HTTP',
  },
  {
    field: 'transport_enabled',
    name: 'TRANSPORT',
  },
  {
    field: 'backend_type',
    name: 'Backend type',
  },
  {
    field: 'backend_configuration',
    name: 'Backend configuration',
    render: (config: object) => renderExpression('Backend configuration', config),
  },
];

const ENABLED_STRING = 'Enabled';
const DISABLED_STRING = 'Disabled';

export function AuthorizationPanel(props: {
  authz: [];
  loading: boolean;
  config: ClientConfigType;
}) {
  const [query, setQuery] = useState<Query | null>(null);
  const domains = keys(props.authz);

  const items = map(domains, function (domain: string) {
    const data = get(props.authz, domain);
    // @ts-ignore
    const backend = data.authorization_backend;
    return {
      domain_name: domain,
      // @ts-ignore
      http_enabled: data.http_enabled ? ENABLED_STRING : DISABLED_STRING,
      // @ts-ignore
      transport_enabled: data.transport_enabled ? ENABLED_STRING : DISABLED_STRING,
      backend_type: backend.type,
      backend_configuration: backend.config,
    };
  });

  const search: EuiSearchBarProps = {
    box: {
      placeholder: 'Search authorization domain',
    },
    onChange: (arg) => {
      setQuery(arg.query);
      return true;
    },
  };

  const headerText = 'Authorization';

  const emptyListMessage = (
    <EuiEmptyPrompt
      title={<h3>No authorization</h3>}
      titleSize="s"
      actions={
        props.config.ui.backend_configurable && (
          <ExternalLinkButton
            href={DocLinks.BackendConfigurationAuthorizationDoc}
            text="Manage via config.yml"
          />
        )
      }
    />
  );

  return (
    <PanelWithHeader
      headerText={headerText}
      headerSubText="After the user authenticates, the security plugin can collect backend roles, such as LDAP groups, from authorization backends. These backends have no execution order; the plugin tries to collect backend roles from all of them."
      helpLink={DocLinks.BackendConfigurationAuthorizationDoc}
      count={Query.execute(query || '', items).length}
    >
      <EuiInMemoryTable
        tableLayout={'auto'}
        columns={columns}
        items={items}
        itemId={'domain_name'}
        pagination={true}
        sorting={true}
        search={search}
        message={showTableStatusMessage(props.loading, items, emptyListMessage)}
      />
    </PanelWithHeader>
  );
}
