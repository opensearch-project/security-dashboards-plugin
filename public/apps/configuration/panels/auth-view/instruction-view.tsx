/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { EuiCode, EuiSpacer, EuiText, EuiTitle } from '@elastic/eui';
import React from 'react';
import { ClientConfigType } from '../../../../types';
import { ExternalLinkButton } from '../../utils/display-utils';
import { DocLinks } from '../../constants';

export function InstructionView(props: { config: ClientConfigType }) {
  return (
    <>
      <EuiTitle size="l">
        <h1>Authentication and authorization</h1>
      </EuiTitle>

      <EuiSpacer size="xxl" />

      <EuiText textAlign="center">
        <h2>You have not set up authentication and authorization</h2>
      </EuiText>

      <EuiText textAlign="center" size="xs" color="subdued" className="instruction-text">
        In order to use Security plugin, you must decide on authentication <EuiCode>authc</EuiCode>{' '}
        and authorization backends <EuiCode>authz</EuiCode>. Use{' '}
        <EuiCode>config/opensearch-security/config.yml</EuiCode> to define how to retrieve and
        verify the user credentials, and how to fetch additional roles from backend system if
        needed.
      </EuiText>

      <EuiSpacer />

      {props.config.ui.backend_configurable && (
        <div style={{ textAlign: 'center' }}>
          <ExternalLinkButton
            fill
            size="s"
            href={DocLinks.BackendConfigurationDoc}
            text="Create config.yml"
          />
        </div>
      )}
    </>
  );
}
