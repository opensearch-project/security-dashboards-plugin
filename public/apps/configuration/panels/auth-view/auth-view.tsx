/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { useContext } from 'react';
import { EuiLoadingContent, EuiPageHeader, EuiSpacer, EuiTitle } from '@elastic/eui';
import { isEmpty } from 'lodash';
import { AuthenticationSequencePanel } from './authentication-sequence-panel';
import { AuthorizationPanel } from './authorization-panel';
import { DocLinks } from '../../constants';
import { AppDependencies } from '../../../types';
import { ExternalLinkButton } from '../../utils/display-utils';
import { getSecurityConfig } from '../../utils/auth-view-utils';
import { InstructionView } from './instruction-view';
import { DataSourceContext } from '../../app-router';
import { SecurityPluginTopNavMenu } from '../../top-nav-menu';
import { AccessErrorComponent } from '../../access-error-component';

export function AuthView(props: AppDependencies) {
  const [authentication, setAuthentication] = React.useState([]);
  const [authorization, setAuthorization] = React.useState([]);
  const [loading, setLoading] = React.useState(false);
  const { dataSource, setDataSource } = useContext(DataSourceContext)!;
  const [errorFlag, setErrorFlag] = React.useState(false);
  const [accessErrorFlag, setAccessErrorFlag] = React.useState(false);

  React.useEffect(() => {
    const fetchData = async () => {
      try {
        setLoading(true);
        const config = await getSecurityConfig(props.coreStart.http, dataSource.id);

        setAuthentication(config.authc);
        setAuthorization(config.authz);
        setErrorFlag(false);
        setAccessErrorFlag(false);
      } catch (e) {
        console.log(e);
        // requests with existing credentials but insufficient permissions result in 403, remote data-source requests with non-existing credentials result in 400
        if (e.response && [400, 403].includes(e.response.status)) {
          setAccessErrorFlag(true);
        }
        setErrorFlag(true);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [props.coreStart.http, dataSource]);

  if (isEmpty(authentication) && !loading) {
    return (
      <>
        <SecurityPluginTopNavMenu
          {...props}
          dataSourcePickerReadOnly={false}
          setDataSource={setDataSource}
          selectedDataSource={dataSource}
        />
        {accessErrorFlag ? (
          <EuiTitle size="l">
            <h1>Authentication and authorization</h1>
          </EuiTitle>
        ) : null}
        {accessErrorFlag ? (
          <AccessErrorComponent
            loading={loading}
            dataSourceLabel={dataSource && dataSource.label}
          />
        ) : (
          <InstructionView config={props.config} />
        )}
      </>
    );
  }

  return (
    <>
      <SecurityPluginTopNavMenu
        {...props}
        dataSourcePickerReadOnly={false}
        setDataSource={setDataSource}
        selectedDataSource={dataSource}
      />
      <EuiPageHeader>
        <EuiTitle size="l">
          <h1>Authentication and authorization</h1>
        </EuiTitle>
        {!loading && !errorFlag && props.config.ui.backend_configurable && (
          <ExternalLinkButton
            href={DocLinks.BackendConfigurationDoc}
            text="Manage via config.yml"
          />
        )}
      </EuiPageHeader>
      {loading ? (
        <EuiLoadingContent />
      ) : accessErrorFlag ? (
        <AccessErrorComponent loading={loading} dataSourceLabel={dataSource && dataSource.label} />
      ) : (
        <>
          {/* @ts-ignore */}
          <AuthenticationSequencePanel authc={authentication} loading={loading} />
          <EuiSpacer size="m" />
          {/* @ts-ignore */}
          <AuthorizationPanel authz={authorization} loading={loading} config={props.config} />
        </>
      )}
    </>
  );
}
