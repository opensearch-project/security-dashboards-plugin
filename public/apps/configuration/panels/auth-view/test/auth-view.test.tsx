/*
 *   Copyright OpenSearch Contributors
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { shallow } from 'enzyme';
import React from 'react';
import { AuthView } from '../auth-view';

jest.mock('react', () => ({
  ...jest.requireActual('react'),
  useContext: jest.fn().mockReturnValue({ dataSource: { id: 'test' }, setDataSource: jest.fn() }), // Mock the useContext hook to return dummy datasource and setdatasource function
}));

// eslint-disable-next-line
const mockAuthViewUtils = require('../../../utils/auth-view-utils');

describe('Auth view', () => {
  const mockCoreStart = {
    http: 1,
  };
  const config = {
    authc: {
      basic_internal_auth_domain: {
        authentication_backend: {
          type: 'intern',
          config: {},
        },
      },
    },
    authz: {
      ldap: {
        http_enabled: true,
      },
    },
  };

  const setState = jest.fn();

  beforeEach(() => {
    jest.spyOn(React, 'useState').mockRestore();
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [[], setState])
      .mockImplementationOnce(() => [[], setState])
      .mockImplementationOnce(() => [false, jest.fn()])
      .mockImplementationOnce(() => [false, jest.fn()])
      .mockImplementationOnce(() => [false, jest.fn()]);
    jest.spyOn(React, 'useEffect').mockImplementationOnce((f) => f());
  });

  it('valid data', (done) => {
    mockAuthViewUtils.getSecurityConfig = jest.fn().mockReturnValue(config);

    shallow(<AuthView coreStart={mockCoreStart as any} navigation={{} as any} />);

    process.nextTick(() => {
      expect(mockAuthViewUtils.getSecurityConfig).toHaveBeenCalledTimes(1);

      expect(setState).toHaveBeenNthCalledWith(1, config.authc);
      expect(setState).toHaveBeenNthCalledWith(2, config.authz);

      done();
    });
  });

  it('return error', (done) => {
    mockAuthViewUtils.getSecurityConfig = jest.fn().mockImplementationOnce(() => {
      throw Error();
    });

    jest.spyOn(console, 'log').mockImplementationOnce(() => {});

    shallow(<AuthView coreStart={mockCoreStart as any} navigation={{} as any} />);

    process.nextTick(() => {
      expect(mockAuthViewUtils.getSecurityConfig).toHaveBeenCalledTimes(1);

      expect(setState).toHaveBeenCalledTimes(0);

      done();
    });
  });

  it('should load access error component', async () => {
    jest.spyOn(React, 'useState').mockRestore();
    jest
      .spyOn(React, 'useState')
      .mockImplementationOnce(() => [[], setState])
      .mockImplementationOnce(() => [[], setState])
      .mockImplementationOnce(() => [false, jest.fn()])
      .mockImplementationOnce(() => [false, jest.fn()])
      .mockImplementationOnce(() => [true, jest.fn()]);
    mockAuthViewUtils.getSecurityConfig = jest
      .fn()
      .mockRejectedValue({ response: { status: 403 } });
    const component = shallow(<AuthView coreStart={mockCoreStart as any} navigation={{} as any} />);
    expect(component).toMatchSnapshot();
  });
});
