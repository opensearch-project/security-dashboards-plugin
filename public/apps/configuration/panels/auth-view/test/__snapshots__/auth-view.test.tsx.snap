// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Auth view should load access error component 1`] = `
<Fragment>
  <Memo()
    coreStart={
      Object {
        "http": 1,
      }
    }
    dataSourcePickerReadOnly={false}
    navigation={Object {}}
    selectedDataSource={
      Object {
        "id": "test",
      }
    }
    setDataSource={[MockFunction]}
  />
  <EuiTitle
    size="l"
  >
    <h1>
      Authentication and authorization
    </h1>
  </EuiTitle>
  <AccessErrorComponent
    loading={false}
  />
</Fragment>
`;
